drivers/serial/uart_nrfx_uarte2.c

```diff
@@ -0,0 +1,1053 @@
+	uint32_t flags = NRFX_UARTE_RX_ENABLE_CONT |
+			 get_keep_fifo_content_flag(dev) |
+			 (is_async_api(dev) ? NRFX_UARTE_RX_ENABLE_STOP_ON_END : 0);
+
+	if (cfg->flags & UARTE_CFG_FLAG_NO_RX) {
+		return -ENOTSUP;
+	}
+
+	if (timeout != SYS_FOREVER_US) {
+		/* Set minimum interval to 3 RTC ticks. 3 is used due to RTC limitation
+		 * which cannot set timeout for next tick. Assuming delay in processing
+		 * 3 instead of 2 is used. Note that lower value would work in a similar
+		 * way but timeouts would always occur later than expected,  most likely
+		 * after ~3 ticks.
+		 */
+		adata->idle_cnt = RX_TIMEOUT_DIV + 1;
+		adata->rx_timeout = K_USEC(timeout / RX_TIMEOUT_DIV);
+	} else {
+		adata->rx_timeout = K_NO_WAIT;
+	}
+
+	atomic_or(&data->flags, UARTE_DATA_FLAG_RX_ENABLED);
+
+	err = nrfx_uarte_rx_buffer_set(nrfx_dev, buf, len);
+	if (err != NRFX_SUCCESS) {
+		return -EIO;
+	}
+
+	err = nrfx_uarte_rx_enable(nrfx_dev, flags);
+	if (err != NRFX_SUCCESS) {
+		return (err == NRFX_ERROR_BUSY) ? -EBUSY : -EIO;
+	}
+
+	return 0;
+}
+
+static int api_rx_buf_rsp(const struct device *dev, uint8_t *buf, size_t len)
+{
+	const nrfx_uarte_t *nrfx_dev = get_nrfx_dev(dev);
+	struct uarte_nrfx_data *data = dev->data;
+	nrfx_err_t err;
+
+	if (!(data->flags & UARTE_DATA_FLAG_RX_ENABLED)) {
+		return -EACCES;
+	}
+
+	err = nrfx_uarte_rx_buffer_set(nrfx_dev, buf, len);
+	switch (err) {
+	case NRFX_SUCCESS:
+		return 0;
+	case NRFX_ERROR_BUSY:
+		return -EBUSY;
+	default:
+		return -EIO;
+	}
+}
+
+static int api_rx_disable(const struct device *dev)
+{
+	struct uarte_nrfx_data *data = dev->data;
+
+	k_timer_stop(&data->async->rx_timer);
+
+	return (nrfx_uarte_rx_abort(get_nrfx_dev(dev), true, false) == NRFX_SUCCESS) ? 0 : -EFAULT;
+}
+
+static int api_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct uarte_nrfx_config *cfg = dev->config;
+	const nrfx_uarte_t *instance = &cfg->nrfx_dev;
+	nrfx_err_t err;
+
+	if (is_int_driven_api(dev)) {
+		return uart_fifo_read(dev, c, 1) == 0 ? -1 : 0;
+	}
+
+	if (is_async_api(dev)) {
+		return -EBUSY;
+	}
+
+	err = nrfx_uarte_rx_ready(instance, NULL);
+	if (err == NRFX_SUCCESS) {
+		uint8_t *rx_byte = cfg->nrfx_config.rx_cache.p_buffer;
+
+		*c = *rx_byte;
+		err = nrfx_uarte_rx_buffer_set(instance, rx_byte, 1);
+		__ASSERT_NO_MSG(err == NRFX_SUCCESS);
+
+		return 0;
+	}
+
+	return -1;
+}
+
+static void api_poll_out(const struct device *dev, unsigned char out_char)
+{
+	const nrfx_uarte_t *nrfx_dev = get_nrfx_dev(dev);
+	nrfx_err_t err;
+
+	do {
+		/* When runtime PM is used we cannot use early return because then
+		 * we have no information when UART is actually done with the
+		 * transmission. It reduces UART performance however, polling in
+		 * general is not power efficient and should be avoided in low
+		 * power applications.
+		 */
+		err = nrfx_uarte_tx(nrfx_dev, &out_char, 1, NRFX_UARTE_TX_EARLY_RETURN);
+		__ASSERT(err != NRFX_ERROR_INVALID_ADDR, "Invalid address of the buffer");
```

Comment:
Shouldn't there be something like `if (err != NRFX_ERROR_BUSY) break;` here to avoid potential hanging when assertions are disabled or when an unexpected error (like `NRFX_ERROR_FORBIDDEN`) occurs?